File: September2022/C++/BinaryTreePruning.cpp
// https://leetcode.com/problems/binary-tree-pruning/
class Solution {
private:
    bool helper(TreeNode* root) {
        if (root == NULL) {
            return false;
        }
        
        // Check left subtree contains a 1
        bool left = helper(root->left);
        // Check right subtree contains a 1
        bool right = helper(root->right);

        // If not then prune the tree
        if (!left) {
            root->left = NULL;
        }
        if (!right) {
            root->right = NULL;
        }
        
        // Return true if the current node, left or right subtree contains a 1
        return root->val == 1 || left || right;
    }
    
public:
    // TC: O(n)
    // SC: O(logn)
    TreeNode* pruneTree(TreeNode* root) {
        return helper(root) ? root : NULL;
    }
};

File: September2022/Java/BinaryTreePruning.java

// @saorav21994
// TC : O(n)
// SC : O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode pruneTree(TreeNode root) {
        // Do a pre order traversal = Parent->Left->Right
        // In bottom up manner keep of removing nodes if they are/become leaf and has value = 0
        
        if (root == null)
            return null;
        root.left = pruneTree(root.left);
        root.right = pruneTree(root.right);
        
        // prune node if its leaf and has value = zero
        if (root.val == 0 && root.left == null && root.right == null) 
            return null;
        
        // else if not leaf or has value 1 return node
        return root;
    }
}
